src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +24,181 @@
 
+  class ExceptionSinkError(Exception): pass
+
+  # All reset_* methods are ~idempotent!
+  @classmethod
+  def reset_log_location(cls, new_log_location=None):
+    """
+    Class state:
+    - Overwrites `cls._log_dir`, `cls._pid_specific_error_fileobj`, and
+      `cls._shared_error_fileobj`.
+    OS state:
+    - May create a new directory.
+    - Overwrites signal handlers for many fatal signals.
+
+    :raises: :class:`ExceptionSink.ExceptionSinkError` if the directory is not writable.
+    """
+    # We could no-op here if the log locations are the same, but there's no reason not to have the
+    # additional safety of re-acquiring file descriptors each time.
+
+    if new_log_location:
+      # Create the directory if possible, or raise if not writable.
+      cls._check_or_create_new_destination(new_log_location)
+    else:
+      # Interpret a None argument to mean to just refresh the file handles for the current pid.
+      new_log_location = cls._log_dir
+
+    pid_specific_error_stream, shared_error_stream = cls._recapture_fatal_error_log_streams(
+      new_log_location)
+
+    # NB: mutate process-global state!
+    # Send a stacktrace to this file if interrupted by a fatal error.
+    faulthandler.enable(file=pid_specific_error_stream, all_threads=True)
+
+    # NB: mutate the class variables!
+    cls._log_dir = new_log_location
+    cls._pid_specific_error_fileobj = pid_specific_error_stream
+    cls._shared_error_fileobj = shared_error_stream
+
   @classmethod
-  def set_destination(cls, dir_path):
-    cls._destination = dir_path
+  def reset_exiter(cls, exiter):
+    """
+    Class state:
+    - Overwrites `cls._exiter`.
+    Python state:
+    - Overwrites sys.excepthook.
+    """
+    assert(isinstance(exiter, Exiter))
+    # NB: mutate the class variables! This is done before mutating the exception hook, because the
+    # uncaught exception handler uses cls._exiter to exit.
+    cls._exiter = exiter
+    # NB: mutate process-global state!
+    sys.excepthook = cls._log_unhandled_exception_and_exit
 
   @classmethod
-  def get_destination(cls):
-    return cls._destination
+  def reset_interactive_output_stream(cls, interactive_output_stream):
+    """
+    Class state:
+    - Overwrites `cls._interactive_output_stream`.
+    OS state:
+    - Overwrites the SIGUSR2 handler.
+
+    This is where the the error message on exit will be printed to as well.
+    """
+    # NB: mutate process-global state!
+    # TODO: chain=True will log tracebacks to previous values of the trace stream as well -- what
+    # happens if those file objects are eventually closed? Does faulthandler just ignore them?
+    # This permits a non-fatal `kill -31 <pants pid>` for stacktrace retrieval.
+    faulthandler.register(signal.SIGUSR2, interactive_output_stream, chain=True)
+
+    # NB: mutate the class variables!
+    # We don't *necessarily* need to keep a reference to this, but we do here for clarity.
```

Comment:
This field is used in `_exit_with_failure`, so this comment isn't accurate. As commented in `_log_unhandled_exception_and_exit`, maybe the Exiter should just hold onto the stream it wants to use to render a failure to the user, and the ExceptionSink should just handle the exception logs?

Reply:
That sounds like a fantastic idea.

Reply:
Broke out #6596!